core-runner/src/builder.rs

```diff
@@ -0,0 +1,97 @@
+use super::runner::{Config, ExtMessage, InitializeProgramInfo, Runner};
+use alloc::vec::*;
+use gear_core::storage::{
+    InMemoryMessageQueue, InMemoryProgramStorage, InMemoryStorage, InMemoryWaitList,
+};
+
+/// Builder for [`Runner`].
+pub struct RunnerBuilder {
+    config: Config,
+    programs: Vec<InitializeProgramInfo>,
+    storage: InMemoryStorage,
+}
+
+impl RunnerBuilder {
+    /// Default [`Runner`].
+    pub fn new() -> RunnerBuilder {
+        RunnerBuilder {
+            config: Config::default(),
+            programs: vec![],
+            storage: InMemoryStorage::default(),
+        }
+    }
+
+    /// Set [`Config`].
+    pub fn config(mut self, config: Config) -> RunnerBuilder {
+        self.config = config;
+        self
+    }
+
+    /// Set [`Program`] to be initialized on build.
+    pub fn program(self, code: Vec<u8>) -> ProgramBuilder {
+        ProgramBuilder::new(self, code)
+    }
+
+    /// Initialize all programs and return [`Runner`].
+    pub fn build(self) -> Runner<InMemoryMessageQueue, InMemoryProgramStorage, InMemoryWaitList> {
+        let mut runner = Runner::new(&self.config, self.storage);
+        for program in self.programs {
+            runner
+                .init_program(program)
+                .expect("failed to init program");
+        }
+        runner
+    }
+}
+
+pub struct ProgramBuilder {
+    runner_builder: RunnerBuilder,
+    code: Vec<u8>,
+    source_id: u64,
+    new_program_id: u64,
+    message: ExtMessage,
+}
+
+impl ProgramBuilder {
+    pub fn new(runner_builder: RunnerBuilder, code: Vec<u8>) -> ProgramBuilder {
+        ProgramBuilder {
+            runner_builder,
+            code,
+            source_id: 1001,
+            new_program_id: 1,
```

Comment:
this probably should use some counter so that if I add to programs without specifying ids, they have different ids

Reply:
> this probably should use some counter so that if I add to programs without specifying ids, they have different ids

how about `1 + runner_builder.programs.len() as u64` in constructor? and same for init messages.